keps/sig-storage/20200212-volume-group.md

```diff
@@ -0,0 +1,1067 @@
+
+Use case 8:
+Application quiesce is time consuming. Some users may not want to do application quiesce very frequently for that reason. For example, a user may want to run weekly backups with application quiesce and nightly backups without application quiesce but with consistency group support which provides crash consistency across all volumes in the group.
+
+### Goals
+
+* Provide an API to manage multiple volumes together in a group.
+* Provide an API to support consistency groups for snapshots, ensuring crash consistency across all volumes in the group.
+* Provide an API to take a snapshot of a group of volumes, not ensuring crash consistency.
+* Provide a design to facilitate volume placement using the group API (To be determined).
+* The group API should be generic and extensible so that it may be used to support other features in the future.
+* A VolumeGroup may potentially be used to support consistency group replication or group replication in the future, but providing design on replication group is not in the scope of this KEP. This can be discussed in the future.
+
+## Proposal for Consistency Groups and Group Snapshots
+
+This proposal introduces new CRDs VolumeGroup, VolumeGroupClass, and VolumeGroupSnapshot.
+
+Create new VolumeGroup can be done in several ways:
+1. Create an empty group first, then create a new PVC with the group name which will add a volume to the already created group.
+2. Create an empty group first, and then add an existing PVC to the group one by one.
+3. Create a new volume group from an existing group snapshot.
+4. Create a new volume group and add existing PVCs matching the label selector to the group.
+5. Non-goal: Create a new empty group and in the same time create new empty PVCs and add to the new group.
+
+Modify an existing VolumeGroup:
+Add new volume or remove existing volume from an existing VolumeGroup. Option 2 for create VolumeGroup above falls into this case.
+
+### Create and Modify VolumeGroup
+
+VolumeGroups can be created and/or modified in several ways as described in the following.
+
+#### Create new PVC and add to the VolumeGroup
+
+* Create a new empty VolumeGroup.
+* Create a new PVC with existing VolumeGroup name. As a result, new PVC is created and added to VolumeGroup. VolumeGroup is modified so Status has this new PVC in PVCList.
+* External-provisioner will be modified so that VolumeGroupName will be passed to the CSI driver when creating a volume.
+
+Only CSI drivers supporting VOLUMEGROUP capability can support the volume group this way.
+When a new PVC is created with the existing VolumeGroup name, the VolumeGroup will be modified and the PVC will be added to PVCList in the Status.
+
```

Comment:
Should the same PVC belong to different groups, i.e., different type of groups or different groups of the same type?
Let storage system to decide whether to support this or not, but we don't prevent it in API or controller directly.